### ISSUE TITLE
Redundant allocations not detected

### ISSUE THREAD

**Issue by vvv** [rust-lang/rust-clippy#9281]
### Summary

[`REDUNDANT_ALLOCATION`] lint should warn about expressions like `Arc<Box<T>>`.  It doesn't.

[`REDUNDANT_ALLOCATION`]: https://github.com/rust-lang/rust-clippy/blob/e17b97c8e02c7de0230466ec36bb9cceef9c774f/clippy_lints/src/types/mod.rs#L190-L213

### Lint Name

REDUNDANT_ALLOCATION

### Reproducer

I tried this code:

```rust
use std::sync::Arc;

fn main() {
    let ab = Arc::new(Box::new(5_u8));
    assert_eq!(**ab, 5);
    
    let _: Box<&Arc<Box<u8>>> = Box::new(&ab);
}
```

`cargo clippy` issued no warnings.

### Version

```text
rustc 1.62.1 (e092d0b6b 2022-07-16)
binary: rustc
commit-hash: e092d0b6b43f2de967af0887873151bb1c0b18d3
commit-date: 2022-07-16
host: x86_64-unknown-linux-gnu
release: 1.62.1
LLVM version: 14.0.5
```

**Comment by lengyijun** [comment]
`redundant_allocation` only lints on field in struct and parameters in functions.

**Comment by vvv** [comment]
Ah, I see.  Indeed.

What about expressions like `Arc<Box<dyn Trait>>` in struct fields?  Shouldn't clippy report them?  (As of now [it doesn't](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e9341b373ee6ca4319a511e84a020c83).)

**Comment by lengyijun** [comment]
Due to https://github.com/rust-lang/rust-clippy/issues/7487

**Comment by dswij** [comment]
We should at least update the documentation for the lint, it currently says that it checks "*anywhere*"

https://github.com/rust-lang/rust-clippy/blob/2038084cf2253b57cf8b405ab000a92b68346f43/clippy_lints/src/types/mod.rs#L191-L193

**Comment by vvv** [comment]
I don't know how to extend the documentation with `Add<Box<dyn Trait>>` caveat without confusing the reader.

https://github.com/rust-lang/rust-clippy/blob/71777465cc71b365ca08db8afd5c25eb99bd3f08/clippy_lints/src/types/mod.rs#L191-L208

TWIMC: https://www.reddit.com/r/rust/comments/jxykva/sizeof_box_dyn_is_weird/
